### ISSUE TITLE
LSTM Method for DHT11 in Indekos

### ISSUE THREAD

**Issue by guafahmi** [guafahmi/Tubes-InternetOfThingsCerdas-S2-MTE-Telkom#1]
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Dropout
from google.colab import files
import io

# 1. Upload File
print("Silakan upload file 'data_sensor_fahmi (2) (1).csv'")
uploaded = files.upload()

# 2. Load Data
file_name = list(uploaded.keys())[0]
df = pd.read_csv(io.BytesIO(uploaded[file_name]))

# 3. Preprocessing Data
# Menggabungkan Tanggal dan Jam menjadi satu kolom Datetime
df['Datetime'] = pd.to_datetime(df['Tanggal'] + ' ' + df['Jam'])
df = df.sort_values('Datetime') # Pastikan data urut waktu

# Mengambil kolom target
data = df[['Suhu (°C)', 'Kelembapan (%)']].values

# Normalisasi data (0 sampai 1) agar LSTM bekerja optimal
scaler = MinMaxScaler(feature_range=(0, 1))
scaled_data = scaler.fit_transform(data)

# Fungsi untuk membuat struktur data sequence (lookback)
def create_sequences(data, lookback):
    X, y = [], []
    for i in range(len(data) - lookback):
        X.append(data[i:(i + lookback)])
        y.append(data[i + lookback])
    return np.array(X), np.array(y)

# Parameter: menggunakan 60 data sebelumnya (2 jam terakhir) untuk prediksi
lookback = 60
X, y = create_sequences(scaled_data, lookback)

# Split data (80% training, 20% testing)
train_size = int(len(X) * 0.8)
X_train, X_test = X[:train_size], X[train_size:]
y_train, y_test = y[:train_size], y[train_size:]

# 4. Membangun Model LSTM
model = Sequential([
    LSTM(50, return_sequences=True, input_shape=(X_train.shape[1], X_train.shape[2])),
    Dropout(0.2),
    LSTM(50, return_sequences=False),
    Dropout(0.2),
    Dense(2) # Output 2: Suhu dan Kelembapan
])

model.compile(optimizer='adam', loss='mean_squared_error')

# 5. Training Model
print("Memulai proses training...")
history = model.fit(X_train, y_train, epochs=20, batch_size=32, validation_data=(X_test, y_test), verbose=1)

# 6. Prediksi 1 Jam ke Depan (30 steps karena interval 2 menit)
last_sequence = scaled_data[-lookback:] # Ambil data terakhir dari dataset
current_batch = last_sequence.reshape((1, lookback, 2))

future_predictions = []
steps_to_predict = 30 # 30 * 2 menit = 60 menit (1 jam)

for i in range(steps_to_predict):
    # Prediksi satu langkah ke depan
    pred = model.predict(current_batch, verbose=0)
    future_predictions.append(pred[0])

    # Update batch untuk menyertakan hasil prediksi terbaru (rolling forecast)
    current_batch = np.append(current_batch[:, 1:, :], [pred], axis=1)

# Inverse transform untuk mengembalikan ke skala asli
future_predictions = scaler.inverse_transform(future_predictions)

# 7. Visualisasi Hasil Prediksi
plt.figure(figsize=(15, 6))

# Plot Suhu
plt.subplot(1, 2, 1)
plt.plot(range(steps_to_predict), future_predictions[:, 0], marker='o', color='red', label='Prediksi Suhu')
plt.title('Prediksi Suhu 1 Jam Kedepan')
plt.xlabel('Menit (Interval 2 Menit)')
plt.ylabel('Suhu (°C)')
plt.grid(True)
plt.legend()

# Plot Kelembapan
plt.subplot(1, 2, 2)
plt.plot(range(steps_to_predict), future_predictions[:, 1], marker='o', color='blue', label='Prediksi Kelembapan')
plt.title('Prediksi Kelembapan 1 Jam Kedepan')
plt.xlabel('Menit (Interval 2 Menit)')
plt.ylabel('Kelembapan (%)')
plt.grid(True)
plt.legend()

plt.tight_layout()
plt.show()

# Menampilkan hasil angka
future_df = pd.DataFrame(future_predictions, columns=['Suhu (Prediksi)', 'Kelembapan (Prediksi)'])
print("\n--- Hasil Prediksi 1 Jam Kedepan ---")
print(future_df)
